Add privacy implantation for logging methods (#641)
microsoft/code-with-engineering-playbook: privacy/recipes/logging.md

        return new_message

```
## Specific tools solutions

### Application Insights

> Application Insights offers telemetry interception in some of the SDKs, that can be done by implementing the `ITelemetryProcessor` interface.

Reviewer: ```suggestion

Application Insights offers telemetry interception in some of the SDKs, that can be done by implementing the `ITelemetryProcessor` interface.
```

> ITelemetryProcessor processes the telemetry information before it is sent to Application Insights, and can be useful in many situations, such as filtering and modifiections. Below is an example of intercepting 'trace' typed telemetry:

Reviewer: ```suggestion
ITelemetryProcessor processes the telemetry information before it is sent to Application Insights, and can be useful in many situations, such as filtering and modifications. Below is an example of intercepting 'trace' typed telemetry:
```

```csharp
using Microsoft.ApplicationInsights.DataContracts;

namespace Example
{
    using Microsoft.ApplicationInsights.Channel;
    using Microsoft.ApplicationInsights.Extensibility;

    internal class RedactTelemetryInitializer : ITelemetryInitializer
    {
        public void Initialize(ITelemetry telemetry)
        {
            var requestTelemetry = telemetry as TraceTelemetry;
            if (requestTelemetry == null) return;
            # redact emails from the message parameter
            requestTelemetry.Message = Regex.Replace(requestTelemetry.Message, @"[^@\s]+@[^@\s]+\.[^@\s]+", "[email removed]");
        }
    }
}
```

### Elastic Stack

>  Elastic Stack (formerly "ELK stack") allows logs interception by Logstash's [filter-plugins](https://www.elastic.co/guide/en/logstash/current/filter-plugins.html). Using some of the existing plugins, like 'mutate', 'alter' and 'prune' might be sufficiant for most cases of deidentifying and redacting PIIs. For a more robust and custimized usecase, a 'ruby' plugin can be used, executing arbitrary Ruby code. Filter plugins also exists in some Logstash alternatives, like [Fluentd](https://docs.fluentd.org/filter) and (Fluent Bit)[https://docs.fluentbit.io/manual/pipeline/filters].

Reviewer: ```suggestion

Elastic Stack (formerly "ELK stack") allows logs interception by Logstash's [filter-plugins](https://www.elastic.co/guide/en/logstash/current/filter-plugins.html). Using some of the existing plugins, like 'mutate', 'alter' and 'prune' might be sufficient for most cases of deidentifying and redacting PIIs. For a more robust and customized use-case, a 'ruby' plugin can be used, executing arbitrary Ruby code. Filter plugins also exists in some Logstash alternatives, like [Fluentd](https://docs.fluentd.org/filter) and (Fluent Bit)[https://docs.fluentbit.io/manual/pipeline/filters].
```

### Presidio

> [Presidio](https://github.com/microsoft/presidio) offers data protection and anonymization API, and can be used alogside to the log interception methods above. Presidio is containerized for REST HTTP API and also can be installed as a python package, to be called from python code. Instead of handeling the anonymization in the application code, both APIs can be used using external calls.

Reviewer: ```suggestion

[Presidio](https://github.com/microsoft/presidio) offers data protection and anonymization API, and can be used alongside to the log interception methods above. Presidio is containerized for REST HTTP API and also can be installed as a python package, to be called from python code. Instead of handling the anonymization in the application code, both APIs can be used using external calls.
```

Elastic Stack, for example, can handle PII redaction using the 'ruby' filter plugin to call Presidio in REST HTTP API, or by calling a python script consuming Presidio as a package:

`logstash.conf`
```ruby

input {
    ...
}

filter {
   ruby {
    code => 'require "open3"
             message = event.get("message")
             # Call a python script triggering Presidio analyzer and anonymizer, and printing the result.
             cmd =  "python /path/to/presidio/anonymization/script.py \"#{message}\""
             # Fetch the script's stdout
             stdin, stdout, stderr = Open3.popen3(cmd)
             # Override message with the anonymized text.
             event.set("message", stdout.read)
             filter_matched(event)'
   }
}

output {
    ...
}
```

> 
> 
> 
> 

Reviewer: ```suggestion

```
